apache2/waf_logging/waf_log_util.cc

```diff
@@ -1,122 +1,269 @@
+        rule_set_versions::VERSION_UNKNOWN;
+}
+
+using rule_ids_vector = std::vector<int>;
+
+static rule_ids_vector current_rule_ids;
+static std::map<rule_set_info, rule_ids_vector> disableable_rule_ids_map;
+
+static const std::string WAF_ACTION_BLOCKED {"Blocked"};
+static const std::string WAF_ACTION_DETECTED {"Detected"};
+static const std::string WAF_ACTION_MATCHED {"Matched"};
+
+static bool is_mandatory_rule(const rule_set_types type, const rule_set_versions version, const int current_id) {
+    const auto contains_current_id = [current_id] (const rule_ids_vector& rule_ids) {
+        return !std::binary_search(std::begin(rule_ids), std::end(rule_ids), current_id);
+    };
+    const auto rules_id_iter = disableable_rule_ids_map.find(std::make_pair(type, version));
+    if (rules_id_iter != std::end(disableable_rule_ids_map)) {
+        return contains_current_id(rules_id_iter->second);
     }
-    
-    if (details_line != NULL) {
-		details_line[strlen(details_line) - 1] = '\0';
+
+    return contains_current_id(current_rule_ids);
+}
+
+static std::string get_json_log_message(const char* resource_id, const char* operation_name, const char* category,
+        const char* instance_id, const char* client_ip, const char* client_port, const char* request_uri,
+        const char* rule_set_type, const char* rule_set_version, const char* rule_id, const char* messages,
+        const int action, const int site, const char* details_messages, const char* details_data,
+        const char* details_file, const char* details_line, const char* hostname, const char* waf_unique_id) {
+
+    using str_view = jsoncons::string_view;
+    const auto make_string_view_strip_quotes = [] (const char* str) -> str_view {
+        if (!str) {
+            return {};
+        }
+        auto length = strlen(str);
+        if (length < 2) {
+            return {};
+        }
+
+        return {str + 1, length - 2};
+    };
+    const str_view rule_id_str = make_string_view_strip_quotes(rule_id);
+    const rule_set_types type = parse_rule_set_type(rule_set_type);
+    const rule_set_versions version = parse_rule_set_version(rule_set_version);
+    const int current_id = std::stoi(static_cast<std::string>(rule_id_str));
+    const bool is_mandatory = is_mandatory_rule(type, version, current_id);
+
+    const str_view action_str =
+        (version != rule_set_versions::VERSION_2_2_9 && !is_mandatory)  ? WAF_ACTION_MATCHED :
+        action == MODSEC_MODE_DETECT                                    ? WAF_ACTION_DETECTED :
+        action == MODSEC_MODE_PREVENT                                   ? WAF_ACTION_BLOCKED :
+        str_view{};
+
+    #define MANDATORY_RULE_MESSAGE ("Mandatory rule. Cannot be disabled. ")
+    #define MANDATORY_RULE_MESSAGE_LENGTH (sizeof(MANDATORY_RULE_MESSAGE) - 1)
+    std::array<char, 1024> mandatory_message {MANDATORY_RULE_MESSAGE};
+    str_view message_str;
+    if (messages) {
```

Comment:
since we already put all the logic into separate functions, why do we still keep message logic here?

Reply:
We could separate it out, agreed.